### ISSUE TITLE
drupgalgap_goto weird behaviour 

### ISSUE THREAD

**Issue by scotthooker** [signalpoint/DrupalGap#76]
I have had to wrap the drupalgap_goto in the following as if I don't the page it goto doesn't render. I've put an alert in to see if it is loading the page and that gets called but none of the page renders if I don't wrap the drupalgap_goto in "    drupalgap.services.drupalgap_system.connect.call" 

  if (drupalgap_user_has_role('anonymous user')) {
    drupalgap.services.drupalgap_system.connect.call({
      'success': function() {
        drupalgap_goto('login');
      }
    });
    return;

  }

**Comment by signalpoint** [comment]
@scotthooker - Is 'login' a custom page you made in your App? If not, you'll want to link to 'user/login' instead to load the default login form.

When the app first loads the drupalgap_deviceready() makes a call to DrupalGap System Connect so you shouldn't have to manually.

I'm able to click the 'Login' link (e.g. drupalgap_goto('user/login')) without problems.

Will you please elaborate a bit more on the issue? I looked at your forked repo, it appears to be behind a bit, please ensure that you have pulled in the latest changes.

Also, it would be helpful if you could commit your code to your forked branch, then reference it here, and I can take a look in more detail, thanks!

**Comment by scotthooker** [comment]
I didn't realise I had forked the repo!  I've reforked and commited in my files. (is there a way to update a forked repo? Do I just pull the latest changes to it?) 

Yes 'login' is a custom page defined in my hook menu

I've updated the repo and commited in my app - not the best code I've ever written and wouldn't normal show it off! Has lot of steps to improve it at the moment! JS isn't my strongest language.

What the app does is if I go to a page that needs me to be login - it does a quick check and if I am not logged in it does a drupalgap go to. (line 358 of bettracks.js inside my bettracks module)

 if (drupalgap_user_has_role('anonymous user')) {
    drupalgap.services.drupalgap_system.connect.call({
      'success': function() {
        drupalgap_goto('login');
      }
    });
    return;

  }

Basically if I don't wrap it in     

drupalgap.services.drupalgap_system.connect.call({
      'success': function() {
     }
    });

then the page that returns just doesn't load

**Comment by signalpoint** [comment]
@scotthooker - thanks for sharing your code, I took a quick glance at it and will check it out in more detail soon.

For starters, you should be able to just do this:

if (drupalgap.user.uid == 0) {
  // do some stuff...
}

Instead of this:

if (drupalgap_user_has_role('anonymous user')) {
  // do some stuff...
}

Unless you need a manual call to the DrupalGap System Connect, but by default the App will perform a DrupalGap System Connect on startup (unless you implement hook_deviceready() and return false).

I'd also like to point out that there is now basic support for jQM page events that you can define in your hook_menu():

http://drupalgap.org/node/116

It still has some bugs that need to be worked out, so you may be able to use that instead of stuff like this:

$('#livescores').on('pageinit', function(event) {
  // do some stuff...
});

Your bettracks_login() function seems strange, I wouldn't recommend putting a pageinit event in there, instead define a pageinit handler function and specify it via hook_menu().

Also, you can make alterations to the default login form with DrupalGap's hook_form_alter() in your DrupalGap module, e.g. bettracks_form_alter (form, form_state) { /\* do some stuff */ }

That way you may be able to get rid of your custom login page.

**Comment by scotthooker** [comment]
It's all in dev status at the moment buddy. So things I've written arn't all final and some stuff I wouldn't normal show to people as its just me playing around! 

All sounds good though so thanks for taking a look!

**Comment by signalpoint** [comment]
No prob, thanks for sharing! I think if you attach your pageinit handler via hook_menu() that should fix the problem. If it doesn't feel free to re-open the issue, thanks!
